### PR TITLE
Add service connection to connect with bazaar instead of broadcastReciever

### PR DESCRIPTION
**Pull request by shayanpourvatan**
In the first step, we used `broadcastReceiver` component to communicate with the bazaar, in this MR we add the service component as the main communication component and set `broadcastReciever` as a second component

### CODE REVIEW

**Review by vahabghadiri — APPROVED** [review]

File: auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        } catch (e: Exception) {
+            onException.invoke()
+        }
+    }
+
+    private fun <T> withServiceInBackground(
+        func: (InAppBazaarAuth) -> T,
+        onException: () -> T
+    ) {
+        val runnable = {
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt:115]: Move this logic to another method in order to DRY principle.

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+internal class ServiceStorageConnection(
+    private val context: Context,
+    private val backgroundThread: BackgroundThread,
+    private val mainThread: MainThread
+) : StorageConnection(context), ServiceConnection {
+
+    private var storageService: InAppBazaarStorage? = null
+    private var connectionThreadSecure: AbortableCountDownLatch? = null
+
+    fun connect(): Boolean {
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:28]: This logic here is similar to `ServiceAuthConnection`, You can handle that in a better way in order to have DRY principles(Don't Repeat Yourself)

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                null
+            }
+        )
+    }
+
+    private fun saveData(connection: InAppBazaarStorage, data: ByteArray): Bundle? {
+        return connection.saveData(context.packageName, data.toBase64())
+    }
+
+    override fun disconnect(context: Context) {
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:115]: Same here (Dry)

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        context.unbindService(this)
+        storageService = null
+        connectionThreadSecure?.abort()
+
+        backgroundThread.interrupt()
+        super.disconnect(context)
+    }
+
+    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:124]: Same Here

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        storageService = InAppBazaarStorage.Stub.asInterface(service)
+        connectionThreadSecure?.countDown()
+    }
+
+    override fun onServiceDisconnected(name: ComponentName?) {
+        storageService = null
+    }
+
+    private fun <T> withService(
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:133]: Same Here (I dont know if we can handle InAppBazaarStorage and Auth in the same way

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageConnection.kt [hunk 6]
```diff
@@ -14,21 +18,69 @@ internal abstract class StorageConnection(private val context: Context) {
     abstract fun saveData(owner: LifecycleOwner?, data: ByteArray, callback: BazaarStorageCallback)
     abstract fun saveDataSync(owner: LifecycleOwner?, data: ByteArray)
 
+    open fun disconnect(context: Context) {
+        storageConnection = null
+    }
+
+    fun getGetSavedDataResponse(extras: Bundle?): BazaarResponse<ByteArray> {
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageConnection.kt:25]: Two get in the name is a little confusing, You can use something like `retriveGetSavedDataResponse`

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageResponseHandler.kt [hunk 7]
```diff
@@ -15,24 +15,29 @@ object StorageResponseHandler {
     private const val IAL_STATUE_FAILED: Int = -1
     private const val IAL_STATUE_DEVELOPER_ERROR: Int = -2
     private const val IAL_STATUE_ACCOUNT_NOT_EXISTS: Int = -3
+    private const val IAL_STATUE_EXTRA_IS_NULL: Int = -4
```
Comment (hamidrezasahraei) [inline auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageResponseHandler.kt:18]: STATUE is wrong (change the above names also)

**Review by hamidrezasahraei — CHANGES_REQUESTED** [review]

File: auth/src/main/aidl/com/farsitel/bazaar/auth/InAppBazaarAuth.aidl [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+* This class used to communicate with the inAppBazaar library to get the account info of the user from Bazaar
+*/
+interface InAppBazaarAuth {
+    /**
+    *  @params: packageName of the thirdParty application
+    *  @Return: Bundle with the following data:
+            ID
+            STATUE
```
[the remark this answers is not in the document]
Comment (shayanpourvatan) [reply auth/src/main/aidl/com/farsitel/bazaar/auth/InAppBazaarAuth.aidl:11]: Done

File: auth/src/main/java/com/farsitel/bazaar/auth/connection/AuthConnection.kt [hunk 9]
```diff
@@ -1,32 +1,67 @@
...
+import com.farsitel.bazaar.thread.BackgroundThread
+import com.farsitel.bazaar.util.InAppLoginLogger
 import com.farsitel.bazaar.thread.MainThread
 
 internal abstract class AuthConnection(private val context: Context) {
 
     abstract fun getLastAccountId(owner: LifecycleOwner?, callback: BazaarSignInCallback)
     abstract fun getLastAccountIdSync(owner: LifecycleOwner?): BazaarResponse<BazaarSignInAccount>?
 
+    open fun disconnect(context: Context) {
```
[the remark this answers is not in the document]
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/auth/connection/AuthConnection.kt:18]: Done

File: auth/src/main/java/com/farsitel/bazaar/auth/connection/AuthResponseHandler.kt [hunk 10]
```diff
@@ -12,22 +12,26 @@ object AuthResponseHandler {
...
 
     private const val ERROR_MESSAGE_UNKNOWN = "unknown error"
+    private const val ERROR_ACCOUNT_NOT_EXISTS = "Account not exists for packageName"
+    private const val ERROR_EXTRA_IS_NULL = "extra data is null"
 
     private const val IAL_STATUE_SUCCESS: Int = 0
     private const val IAL_STATUE_FAILED: Int = -1
     private const val IAL_STATUE_DEVELOPER_ERROR: Int = -2
     private const val IAL_STATUE_ACCOUNT_NOT_EXISTS: Int = -3
+    private const val IAL_STATUE_EXTRA_IS_NULL: Int = -4
```
[the remark this answers is not in the document]
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/auth/connection/AuthResponseHandler.kt:22]: Done

File: auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+    private fun getLastAccountBundleFromService(
+        connection: InAppBazaarAuth
+    ): BazaarResponse<BazaarSignInAccount> {
+        val bundle = connection.getLastAccountId(context.packageName)
+        return getLastAccountResponse(bundle)
+    }
+
+    override fun disconnect(context: Context) {
+        context.unbindService(this)
```
[the remark this answers is not in the document]
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt:58]: Done

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt (hunk 3 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:115]: Done

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageConnection.kt (hunk 6 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageConnection.kt:25]: Done

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageResponseHandler.kt (hunk 7 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/storage/connection/StorageResponseHandler.kt:18]: Done

File: auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt (hunk 1 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/auth/connection/ServiceAuthConnection.kt:115]: Done

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt (hunk 5 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt:133]: Done

File: auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt (hunk 4 above)
Comment (shayanpourvatan) [reply auth/src/main/java/com/farsitel/bazaar/storage/connection/ServiceStorageConnection.kt]: Done

**Comment by hamidrezasahraei** [comment]
Now LGTM
@shayanpourvatan

**Pull request merged**